orderbook/src/api/validation.rs

```diff
@@ -0,0 +1,353 @@
+            ));
+        }
+        if !matches!(
+            order.sell_token_balance,
+            SellTokenSource::Erc20 | SellTokenSource::External
+        ) {
+            return Err(ValidationError::UnsupportedSellTokenSource(
+                order.sell_token_balance,
+            ));
+        }
+
+        if order.valid_to
+            < shared::time::now_in_epoch_seconds() + self.min_order_validity_period.as_secs() as u32
+        {
+            return Err(ValidationError::InsufficientValidTo);
+        }
+
+        if has_same_buy_and_sell_token(&order, &self.native_token) {
+            return Err(ValidationError::SameBuyAndSellToken);
+        }
+
+        if order.buy_token == BUY_ETH_ADDRESS {
+            let code_size = self
+                .code_fetcher
+                .code_size(order.receiver)
+                .await
+                .map_err(ValidationError::Other)?;
+            if code_size != 0 {
+                return Err(ValidationError::TransferEthToContract);
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Returns true if the orders have same buy and sell tokens.
+///
+/// This also checks for orders selling wrapped native token for native token.
+fn has_same_buy_and_sell_token(order: &PreOrderData, native_token: &WETH9) -> bool {
+    order.sell_token == order.buy_token
+        || (order.sell_token == native_token.address() && order.buy_token == BUY_ETH_ADDRESS)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use anyhow::anyhow;
+    use shared::dummy_contract;
+    use shared::web3_traits::MockCodeFetching;
+
+    #[test]
+    fn detects_orders_with_same_buy_and_sell_token() {
+        let native_token = dummy_contract!(WETH9, [0xef; 20]);
+        assert!(has_same_buy_and_sell_token(
+            &PreOrderData {
+                sell_token: H160([0x01; 20]),
+                buy_token: H160([0x01; 20]),
+                ..Default::default()
+            },
+            &native_token,
+        ));
+        assert!(has_same_buy_and_sell_token(
+            &PreOrderData {
+                sell_token: native_token.address(),
+                buy_token: BUY_ETH_ADDRESS,
+                ..Default::default()
+            },
+            &native_token,
+        ));
+
+        assert!(!has_same_buy_and_sell_token(
+            &PreOrderData {
+                sell_token: H160([0x01; 20]),
+                buy_token: H160([0x02; 20]),
+                ..Default::default()
+            },
+            &native_token,
+        ));
+        // Sell token set to 0xeee...eee has no special meaning, so it isn't
+        // considered buying and selling the same token.
+        assert!(!has_same_buy_and_sell_token(
+            &PreOrderData {
+                sell_token: BUY_ETH_ADDRESS,
+                buy_token: native_token.address(),
+                ..Default::default()
+            },
+            &native_token,
+        ));
+    }
+
+    #[tokio::test]
+    async fn validate_partial_order_err() {
```

Comment:
One underrated feature of this refactor is the ability to test order validation better.

Reply:
Perhaps in a follow up PR, I can make a similar method for the other parts of order validation.